uobtheatre/payments/models.py

```diff
@@ -281,38 +293,92 @@ def refund(self, refund_provider: RefundProvider = None):
         refund_provider.refund(self)
 
 
-TOTAL_PROVIDER_FEE = Coalesce(
-    Sum(
-        "provider_fee",
-    ),
-    0,
-)
-NET_TOTAL = Sum("value")
-NET_CARD_TOTAL = Sum("value", filter=(~Q(provider_name=Cash.name)))
-TOTAL_SALES = Sum("value", filter=Q(type=Transaction.Type.PAYMENT))
-TOTAL_CARD_SALES = Sum(
-    "value", filter=(~Q(provider_name=Cash.name) & Q(type=Transaction.Type.PAYMENT))
-)
-TOTAL_REFUNDS = Sum("value", filter=Q(type=Transaction.Type.REFUND))
-TOTAL_CARD_REFUNDS = Sum(
-    "value", filter=(~Q(provider_name=Cash.name) & Q(type=Transaction.Type.REFUND))
-)
-APP_FEE = Coalesce(Sum("app_fee"), 0)
-
-SALE_BREAKDOWN_ANNOTATIONS: dict[str, Any] = {
-    # Gross Income
-    "net_income": NET_TOTAL,
-    "net_card_income": NET_CARD_TOTAL,
-    # Total Purchases / Sales
-    "total_sales": TOTAL_SALES,
-    "total_card_sales": TOTAL_CARD_SALES,
-    # Total Refunds
-    "total_refunds": TOTAL_REFUNDS,
-    "total_card_refunds": TOTAL_CARD_REFUNDS,
-    # Gross Amount charged by the payment provider (square) for these payments
-    "provider_payment_value": TOTAL_PROVIDER_FEE,
-    # Amount we take from net payment - provider cut
-    "app_payment_value": APP_FEE - TOTAL_PROVIDER_FEE,
-    "society_transfer_value": NET_CARD_TOTAL - APP_FEE,
-    "society_revenue": NET_TOTAL - APP_FEE,
-}
+class SalesBreakdown:
+    """
+    Class representing the sales breakdown for a given transaction query set
+    """
+
+    class Enums(Enum):
+        """
+        The available breakdowns.
+
+        Each enum variant is assigned to the expression used to calculate its
+        value.
+        """
+
+        PROVIDER_PAYMENT_VALUE = Coalesce(
+            Sum(
+                "provider_fee",
+            ),
+            0,
+        )
+        NET_TRANSACTIONS = Sum("value")
+        NET_CARD_TRANSACTIONS = Sum("value", filter=(~Q(provider_name=Cash.name)))
+        TOTAL_PAYMENTS = Sum("value", filter=Q(type=Transaction.Type.PAYMENT))
+        TOTAL_CARD_PAYMENTS = Sum(
+            "value",
+            filter=(~Q(provider_name=Cash.name) & Q(type=Transaction.Type.PAYMENT)),
+        )
+        TOTAL_REFUNDS = Sum("value", filter=Q(type=Transaction.Type.REFUND))
+        TOTAL_CARD_REFUNDS = Sum(
+            "value",
+            filter=(~Q(provider_name=Cash.name) & Q(type=Transaction.Type.REFUND)),
+        )
+        APP_FEE = Coalesce(Sum("app_fee"), 0)
+
+        APP_PAYMENT_VALUE = APP_FEE - PROVIDER_PAYMENT_VALUE
+        SOCIETY_TRANSFER_VALUE = NET_CARD_TRANSACTIONS - APP_FEE
+        SOCIETY_REVENUE = NET_TRANSACTIONS - APP_FEE
+
+        @property
+        def key(self):
+            return self.name.lower()
+
+    def __init__(self, transaction_qs: TransactionQuerySet) -> None:
+        super().__init__()
+        self.transaction_qs = transaction_qs
+
+    @property
+    def total_payments(self) -> int:
```

Comment:
Can we create a ticket so that the query smashes this get_sales_breakdown into a single query is multiple sales breakdowns are required

Reply:
Go for it